flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -260,17 +258,28 @@ protected StreamTask(
 			@Nullable TimerService timerService,
 			Thread.UncaughtExceptionHandler uncaughtExceptionHandler,
 			StreamTaskActionExecutor actionExecutor,
-			TaskMailbox mailbox) {
+			TaskMailbox mailbox) throws Exception {
 
 		super(environment);
 
-		this.timerService = timerService;
-		this.uncaughtExceptionHandler = Preconditions.checkNotNull(uncaughtExceptionHandler);
 		this.configuration = new StreamConfig(getTaskConfiguration());
 		this.recordWriter = createRecordWriterDelegate(configuration, environment);
 		this.actionExecutor = Preconditions.checkNotNull(actionExecutor);
 		this.mailboxProcessor = new MailboxProcessor(this::processInput, mailbox, actionExecutor);
 		this.asyncExceptionHandler = new StreamTaskAsyncExceptionHandler(environment);
+		this.asyncOperationsThreadPool = Executors.newCachedThreadPool(
+			new ExecutorThreadFactory("AsyncOperations", uncaughtExceptionHandler));
+
+		this.stateBackend = createStateBackend();
+		this.checkpointStorage = stateBackend.createCheckpointStorage(getEnvironment().getJobID());
+
+		// if the clock is not already set, then assign a default TimeServiceProvider
+		if (timerService == null) {
+			ThreadFactory timerThreadFactory = new DispatcherThreadFactory(TRIGGER_THREAD_GROUP, "Time Trigger for " + getName());
+			this.timerService = new SystemProcessingTimeService(this::handleTimerException, timerThreadFactory);
+		} else {
```

Comment:
I'm a bit concerned about concurrency here: we publish a reference to this object before it's being fully constructed.
So the other thread can read null as `asyncExceptionHandler` value.
But now this field is final, so the preceding write to it should be already visible.

If this is correct, I think we need to put a warning here and near `final` modifier that changing them can cause concurrency problems.

Reply:
I did not get your point here. You referred to the concurrent issue for `timeService` or `asyncExceptionHandler`? And which kind of thread might touch this object before the `StreamTask` is fully constructed?

Reply:
I meant that the timer thread could access `asyncExceptionHandler` variable.
So, if someone later reorders it like this:
```
StreamTask(){
  this.timerService = new SystemProcessingTimeService(this::handleTimerException, timerThreadFactory);
  this.asyncExceptionHandler = new StreamTaskAsyncExceptionHandler(environment);
}
...
/// schedule task
```
and removes `final` from `asyncExceptionHandler` then `timerService` thread can see `null` even **after** it's assigned.

Reply:
Hmm, thanks for the explanations and I got your point now.

I also found that there exists another variable to prevent this case happen.  `StreamTask#isRunning` is set true after `beforeInvoke` done, and the `asyncExceptionHandler` would be only used by `timeService` after `isRunning` true. So it can eliminate our concern. :)